### ISSUE TITLE
[HOLD for payment 2024-07-10] [$250] [CRITICAL] [UX Reliability] When deeplinking to expense report from email empty screen is shown

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#42753]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @danielrvidal 
**Slack conversation:**

## Action Performed:

1. [daniel@expensify.com](mailto:daniel@expensify.com) submitted [danrvidal+upgradetest7@gmail.com](mailto:danrvidal+upgradetest7@gmail.com) an IOU.
2. [danrvidal+upgradetest7@gmail.com](mailto:danrvidal+upgradetest7@gmail.com) got the email and clicked Pay , is taken to the sign in page, gets the magic code and logs in
3. [danrvidal+upgradetest7@gmail.com](mailto:danrvidal+upgradetest7@gmail.com) is taken to directly to the expense.
4. The subheader is not loading right away and LHN chats are not loading right away.
5. After letting it load, I click into the DM and there is nothing showing.
  a. I let it load for a bit, nothing.
  b. I refresh and see the expense show up.

This makes for a confusing experience. As a side note, we’ll be updating the flow so we don’t send users to the expense directly but rather the DM ([GH](https://github.com/Expensify/App/issues/42321)) but I still experienced this bug so it felt like we should update it.

## Expected Result:
The DM chat should load immediately as well as the chats should be available in LHN first, we should wait for the openApp response and show the skeleton if its not available

## Actual Result:
- Subheader not showing when I’m brought into a report for the first time as a new user, and the DM the IOU sent to me is not loading.
- The DM chat is showing empty screen

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

In slack here https://expensify.slack.com/archives/C05LX9D6E07/p1716940223060999?thread_ts=1716939884.285849&cid=C05LX9D6E07

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0167ada75a878e8b10</li>
        <li>Upwork Job ID: 1795785341708906496</li>
        <li>Last Price Increase: 2024-05-29</li>
<li>Automatic offers: </li>
<ul>
<li>c3024 | Reviewer | 102542821</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by Skakruk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Subheader not showing when user brought into a report for the first time as a new user, and the DM the IOU sent to me is not loading. 
Also reproducible for any user, who received IOU link, on the first app load (after clearing DB) when the user navigates to IOU report directly.

### What is the root cause of that problem?

On the first app load, we're making OpenApp request which blocks any other updates to Onyx, until it finishes.
https://github.com/Expensify/App/blob/15600b0d29b1bb745e0c2ea2cf7020bf6e240e96/src/libs/actions/OnyxUpdateManager/index.ts#L74-L80
Since OpenReport request is established before the OpenApp, the Report data is skipped and not saved into IndexedDB.

### What changes do you think we should make in order to solve the problem?

We need to delay OpenReport request until OpenApp is made.

In `ReportScreen.tsx` we add 
```
const [isLoadingApp] = useOnyx(ONYXKEYS.IS_LOADING_APP);
```
 and add checks for `isLoadingApp` in a few places:

1. don't make `fetchReportIfNeeded()` call https://github.com/Expensify/App/blob/21d7884c021fd9f8913d4f317d7d72bfdac9651c/src/pages/home/ReportScreen.tsx#L472-L480
2. for showing skeleton while the app is still loading https://github.com/Expensify/App/blob/21d7884c021fd9f8913d4f317d7d72bfdac9651c/src/pages/home/ReportScreen.tsx#L377
3. to condition `shouldShowNotFoundPage` for not showing "Report not found" https://github.com/Expensify/App/blob/21d7884c021fd9f8913d4f317d7d72bfdac9651c/src/pages/home/ReportScreen.tsx#L388

Also add condition 
```
if (report.reportID === reportIDFromRoute) {
    return;
}
```
 to https://github.com/Expensify/App/blob/21d7884c021fd9f8913d4f317d7d72bfdac9651c/src/pages/home/ReportScreen.tsx#L417 to make sure we don't make a duplicate request for already loaded report.


https://github.com/Expensify/App/assets/1768919/239b517d-c79b-4a56-959a-bb80e27d5143

**Comment by arosiclair** [comment]
I can manage this. Looks like we're having @kacper-mikolajczak take a look

**Comment by kacper-mikolajczak** [comment]
Hi! While trying to reproduce the issue I've gotten the infinite loading spinner on LHN with error message for the report:

https://github.com/Expensify/App/assets/62747088/b247b91c-41ac-4b85-820a-596990fb63a6

It hangs even after `OpenReport` call has ended. Is it a known issue?

**Comment by Skakruk** [comment]
@kacper-mikolajczak I got the same issue, and it also related to requests made prior the `OpenApp`

**Comment by kacper-mikolajczak** [comment]
Understandable, I've tested your approach and it works fine. One thing that we could improve is making a `OpenReport` request in parallel to `OpenApp` but as you said, such change would not be easy with current architecture.

@mountiny I think it's best @Skakruk takes over the PR as he already proposed a solid solution.

**Comment by kacper-mikolajczak** [comment]
What also I've noticed while testing, we are making an `OpenReport` request even for user who was not logged in. Is it intended behaviour?

https://github.com/Expensify/App/assets/62747088/62dcf047-ed9b-4794-b9c4-538f7950f076

**Comment by Skakruk** [comment]
appears to be:
https://github.com/Skakruk/ExpensifyApp/blob/b4c7110dd7a6976c492aedd8c9a660e0076b3aad/src/libs/actions/Report.ts#L2487-L2489

**Comment by arosiclair** [comment]
>@mountiny What do you think about @Skakruk taking over the PR as he already proposed a solid solution?

@Skakruk's proposal looks promising to me as well. Let's implement it 👍🏽

**Comment by arosiclair** [comment]
@Skakruk actually this is still up for discussion please hold.

**Comment by arosiclair** [comment]
Alright, upon some further discussion we realized this requires a backend fix to address. The root problem is that the subtitle in the header needs data from the parent report to be populated ([that's done here](https://github.com/Expensify/App/blob/a51a245ac4a3f4cae0635e1564d530b6c186e3c4/src/libs/ReportUtils.ts#L3303)) and `OpenReport` does not return that data. So we just need to include parent report data in the `OpenReport` response.

I'll take this internal and post a PR to implement that.

**Comment by Skakruk** [comment]
That’s strange, from the logging, I noticed that OpenApp is blocking and
skipping any merges in Onyx until is processed. And changes to api might
not fix the issue.

**Comment by arosiclair** [comment]
To clarify, your RCA wasn't accurate. The log you're seeing [here](https://github.com/Expensify/App/blob/15600b0d29b1bb745e0c2ea2cf7020bf6e240e96/src/libs/actions/OnyxUpdateManager/index.ts#L74-L80) is referring to Onyx updates from Pusher not from API commands. API commands like `OpenReport` can merge onyx updates without issue before `OpenApp` completes.

**Comment by Skakruk** [comment]
From what I traced, API calls are saving data in Onyx via [SaveResponseInOnyx](https://github.com/Expensify/App/blob/6106eca7aa7792321513e56a4b8d9f1bb02fb3a8/src/libs/Middleware/SaveResponseInOnyx.ts#L10), which calls [saveUpdateInformation](https://github.com/Expensify/App/blob/a51a245ac4a3f4cae0635e1564d530b6c186e3c4/src/libs/actions/OnyxUpdates.ts#L121). Inside of it, there is `Onyx.set(ONYXKEYS.ONYX_UPDATES_FROM_SERVER, updateParams);` 
We have global listener https://github.com/Expensify/App/blob/15600b0d29b1bb745e0c2ea2cf7020bf6e240e96/src/libs/actions/OnyxUpdateManager/index.ts#L147-L152 

Here it has callback `handleOnyxUpdateGap` which is referred to in the first place, that is blocking saving API response data into Onyx and IDB as a consequence.

On first app load (when OpenApp is called):

<img width="1439" alt="Screenshot 2024-06-01 at 10 34 42" src="https://github.com/Expensify/App/assets/1768919/bf2e67b3-bfa2-4adb-84fc-a1ae9a918832">


After page refresh (when OpenApp is not called):

<img width="1440" alt="Screenshot 2024-06-01 at 10 35 16" src="https://github.com/Expensify/App/assets/1768919/ad8e6995-68b3-48b4-bd32-69daa4e0bac0">

**Comment by c3024** [comment]
I tried to repro the bug on dev a few times and this is what I have been seeing

>3. [danrvidal+upgradetest7@gmail.com](mailto:danrvidal+upgradetest7@gmail.com) is taken to directly to the expense.

I am not taken to the expense but seeing the not found page similar to [this](https://github.com/Expensify/App/issues/42753#issuecomment-2143343709)

>4. The subheader is not loading right away and LHN chats are not loading right away.

LHN shows loading while the chats are loading with center pane showing the not found page.
>5. After letting it load, I click into the DM and there is nothing showing.
a. I let it load for a bit, nothing.

I think this is due to the issues in comment linking. I saw this a few times in other places as well.


https://github.com/Expensify/App/assets/102477862/46e1840c-c23a-4b7e-b45e-8848d9e11734

**Comment by arosiclair** [comment]
With a 10 second delay added to `OpenApp` I can also reproduce the Not Found screen. Some findings:

- Daniel's [original bug](https://expensify.slack.com/archives/C05LX9D6E07/p1716940223060999?thread_ts=1716939884.285849&cid=C05LX9D6E07) is a little different from what I'm reproducing. You can see that his report clearly loads _before_ `OpenApp` finishes. This should be impossible per this [breakdown](https://github.com/Expensify/App/issues/42753#issuecomment-2143343709)
- Adding thread ancestry to `OpenReport`'s response helps populate the header but doesn't fix the NotFound page
- Preventing the NotFound page from appearing while `ONYXKEYS.IS_LOADING_APP` is true helps, but leaves a permanent skeleton loading page

This is very difficult to debug. @Skakruk's RCA actually seems accurate but the solution is a bit of a workaround. This problem of dropping Onyx updates while `OpenApp` is running is similar to the problem of dropping updates while we handle gaps. We fixed that with the deferred updates queue, so I think we should probably use the queue again in this scenario. I'm going to test queuing `OpenReport`'s updates and see if that fixes the issue tomorrow.

**Comment by arosiclair** [comment]
Alright the deferred updates approach was very promising but ultimately doesn't work. The root problem is that we're performing a write command before OpenApp finishes and dropping the update. If we defer updates for the write command, we still end up ignoring them because the `lastUpdateID` from `OpenApp` will already include the `lastUpdateID` for the write command. 

For example if the `lastUpdateID` for `OpenReport` is 50, `OpenApp` will definitely return a `lastUpdateID` >= 50. Since the `clientLastUpdateID` is now >= 50, we'll ignore the deferred `OpenReport` update (even when OpenApp doesn't include that report data).

So I think the best path forward is to block `OpenReport` on `OpenApp`. This might be a recurring problem for any write command you can deep link to, but for now `OpenReport` is the only problem I'm aware of so we can just patch that directly in the `ReportScreen`.

**Comment by Skakruk** [comment]
@arosiclair so what are the next actions?

**Comment by arosiclair** [comment]
> @arosiclair so what are the next actions?

I'm drafting and testing changes both on the backend and frontend. The frontend changes are similar to your proposal so you'll get payment for that assuming everything pans out well.

**Comment by arosiclair** [comment]
To recap I'm solving two problems

1. `OpenReport` doesn't return parent report data for money requests
2. Deep linking to a report while signed out fails to correctly load the report

I posted https://github.com/Expensify/Auth/pull/11174 to solve 1️⃣
- This fixes the issue
- I removed a restriction on getting thread ancestry in DMs for this. This seemed to reveal a bug where clicking on parent comments from a thread leads to a not found page (the reason why it was restricted in the first place). I need to take a closer look at this before moving forward.

I posted https://github.com/Expensify/App/pull/43307 to solve 2️⃣
- This fixes the issue
- Seems to be causing repeated unnecessary calls to `OpenReport`. I need to look into this before posting the PR.

**Comment by johncschuster** [comment]
Thanks for that great update, @arosiclair!

**Comment by johncschuster** [comment]
Thanks for that great update, @arosiclair!

**Comment by arosiclair** [comment]
Retested and addressed issues in https://github.com/Expensify/Auth/pull/11174. It's approved and ready to be merged as soon as we get approval from the fire room [here](https://expensify.slack.com/archives/C0714QF3A1Z/p1718053511702609).

No extra progress on problem 2️⃣ today. I'll continue work on that tomorrow.

**Comment by arosiclair** [comment]
EOD update:
- https://github.com/Expensify/Auth/pull/11174 is deployed to prod. 1️⃣ is done.
- On `main`, deep linking while signed out is even more broken now than before 🙃
  - After signing out, deep linking to a money request and signing in, `OpenApp` doesn't run and the app never loads
  - This is not reproducible on staging atm
  - The root of this problem is a call to `OpenReport` before `OpenApp`. An Onyx update gap with the response is detected which pauses the `SequentialQueue` and prevents `OpenApp` from running
  - I'm not sure what new change specifically is causing this

I'm now thinking [this PR](https://github.com/Expensify/App/pull/43307) might not be the way to go for solving problem 2️⃣.
- To recap, it attempts to fix deep linking by blocking `OpenReport` until `OpenApp` completes
- This doesn't work reliably because we call these API commands when the `AuthScreens` and `ReportScreen` components mount. Since the `ReportScreen` is a descendent of `AuthScreens`, it will mount first and will not know if the app is being initialized.

Some ideas on a better solution
- Defer any write commands that are queued before `OpenApp`. This would probably need to be done at the [API](https://github.com/Expensify/App/blob/6eb672a1736afbc619e5077885ee2b397b5eaca3/src/libs/API/index.ts) or [SequentialQueue](https://github.com/Expensify/App/blob/4f8ec01a36a821009c2c2b473669393aa1c7a2d8/src/libs/Network/SequentialQueue.ts) level
- Add some kind of interstitial "setup/initializing" page between the SignInPage and the LHN. The setup page would completely process `OpenApp` before navigating to the LHN and the initial `ReportScreen`

Unfortunately, I'm going OOO for the rest of the week. I'll be back on Monday to keep working on this if no progress is made.

**Comment by johncschuster** [comment]
Thanks for the update, @arosiclair! I'll keep Melvin at bay while you're OOO.

**Comment by johncschuster** [comment]
Yes, Melvin, Andrew is working on it.

**Comment by johncschuster** [comment]
Pre-emptive bump while @arosiclair is OOO 👍

**Comment by johncschuster** [comment]
And one more before the weekend 👍

**Comment by arosiclair** [comment]
Since this is critical functionality and https://github.com/Expensify/App/pull/43307 does work, I'm moving forward with it so we can get the fix out ASAP. I'll create a follow up to look into some of the better ideas I mentioned later. ETA on the PR is EOW.

**Comment by johncschuster** [comment]
@arosiclair how's this one going?

**Comment by arosiclair** [comment]
PR just hit staging [yesterday](https://github.com/Expensify/App/pull/43307#issuecomment-2189879481).

**Comment by johncschuster** [comment]
How's this one going, @arosiclair?

**Comment by arosiclair** [comment]
We're just waiting on the prod deploy

**Comment by johncschuster** [comment]
Acknowledged! I will issue payment after the regression window 👍

**Comment by johncschuster** [comment]
@c3024 can you complete the BZ Checklist above when you get a moment? Thank you!

**Comment by c3024** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
 
* [x]  [@c3024] The PR that introduced the bug has been identified. Link to the PR: No specific PR can be held responsible for this bug.
* [x]  [@c3024] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: NA
* [x]  [@c3024] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Not started as this bug cannot be caught earlier.
* [x]  [@c3024] Determine if we should create a regression test for this bug. Yes
* [x]  [@c3024] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. [Test here in the author checklist Tests](https://github.com/Expensify/App/pull/43307#issue-2340756444)

> 1. Log into NewDot as user A
> 2. Verify the app and the initial report load without issue
> 3. Open a DM to user B
> 4. Submit an expense in the DM
> 5. Open the report and copy the link
> 6. Sign out
> 7. Open the link and sign in as user B
> 8. Verify
>  - the app loads
>  - the expense report loads
>  - the header includes a link to the parent report
>

**Comment by johncschuster** [comment]
Payment has been issued. Thanks for your contributions! 🎉
